### ISSUE TITLE
Changing number of AZs of VPC fails to update due to CIDR conflict

### ISSUE THREAD

**Issue by kahkeng** [pulumi/pulumi-awsx#520]
Started with the following script:

```
import * as awsx from "@pulumi/awsx";

// Create a VPC for our cluster.
const vpc = new awsx.ec2.Vpc("pulumi-change-az-vpc", {
    cidrBlock: "10.1.0.0/16",
    numberOfAvailabilityZones: 4,
});
```
After running `up`, I decide to change to 2 AZs instead.
```
    numberOfAvailabilityZones: 2,
```

Running `preview` doesn't report any issues, but running `up` with the above change fails with the following:
```
Updating (dev):
     Type                     Name                            Status                   Info
     pulumi:pulumi:Stack      pulumi_change_az-dev            **failed**               1 error
     └─ awsx:x:ec2:Vpc        pulumi-change-az-vpc
        ├─ awsx:x:ec2:Subnet  pulumi-change-az-vpc-public-0
 +-     │  └─ aws:ec2:Subnet  pulumi-change-az-vpc-public-0   **replacing failed**     [diff: ~cidrBlock]; 1 error
        ├─ awsx:x:ec2:Subnet  pulumi-change-az-vpc-public-1
 +-     │  └─ aws:ec2:Subnet  pulumi-change-az-vpc-public-1   **replacing failed**     [diff: ~cidrBlock]; 1 error
        ├─ awsx:x:ec2:Subnet  pulumi-change-az-vpc-private-1
 +-     │  └─ aws:ec2:Subnet  pulumi-change-az-vpc-private-1  **replacing failed**     [diff: ~cidrBlock]; 1 error
        └─ awsx:x:ec2:Subnet  pulumi-change-az-vpc-private-0
 +-        └─ aws:ec2:Subnet  pulumi-change-az-vpc-private-0  **replacing failed**     [diff: ~cidrBlock]; 1 error

Diagnostics:
  aws:ec2:Subnet (pulumi-change-az-vpc-private-1):
    error: Error creating subnet: InvalidSubnet.Conflict: The CIDR '10.1.192.0/18' conflicts with another subnet
        status code: 400, request id: aac27ce2-3aec-417d-a0a8-7f0216421d73

  pulumi:pulumi:Stack (pulumi_change_az-dev):
    error: update failed

  aws:ec2:Subnet (pulumi-change-az-vpc-private-0):
    error: Error creating subnet: InvalidSubnet.Conflict: The CIDR '10.1.128.0/18' conflicts with another subnet
        status code: 400, request id: 5d679d9d-654d-4711-a3b6-44cbb9c9897c

  aws:ec2:Subnet (pulumi-change-az-vpc-public-0):
    error: Error creating subnet: InvalidSubnet.Conflict: The CIDR '10.1.0.0/18' conflicts with another subnet
        status code: 400, request id: b5e6fdc9-4097-4e05-80ac-b6369ebe3a19

  aws:ec2:Subnet (pulumi-change-az-vpc-public-1):
    error: Error creating subnet: InvalidSubnet.Conflict: The CIDR '10.1.64.0/18' conflicts with another subnet
        status code: 400, request id: 7c98d7e9-fc27-4199-a625-a5e34378cdef
```

Perhaps my expectations are too strict but I would have preferred that any potential errors would show up during the `preview` phase instead of `up` phase, so there is less risk of misconfiguration/inconsistently deployed states.

**Comment by leezen** [comment]
Does this work when you set `{ deleteBeforeReplace: true }` on the `vpc` resource? By default we attempt to create before delete on a replace. But in this case, you'll run into a conflict with the existing CIDR blocks that were already allocated.

I agree you'd prefer to see the violations on the preview. At the moment, there's no logic to check if the resulting semantics of the configuration will be accepted by the provider's API.

**Comment by kahkeng** [comment]
It looks like `deleteBeforeReplace` only works on `aws` but not `awsx` (seems like the 3rd input is of type `ComponentResourceOptions` instead of `CustomResourceOptions`). Is there a way to specify this option for `awsx` components?

**Comment by leezen** [comment]
You could do this via `transformations` -- however, I just tried this and it sadly doesn't work because of the logic in `awsx`

It doesn't actually model the relationships between the subnets well. So, in order to free up the block, it may not only need to replace, but delete some existing resources. Pulumi always attempts to create (or replace) before general deletes. Because the relationship isn't modeled, the deletes don't take place and the CIDR blocks are not necessarily freed up to make this work. I'll leave this open as a bug for the library.

**Comment by mdhornet90** [comment]
When is this being addressed? I basically have to blow away my entire stack to work around this.

**Comment by richmeij** [comment]
I also have a usecase where I changed the name of my subnets, and now I cant run `pulumi up` because of the fact that the CIDR blocks didnt change, and it errors out with: `error creating EC2 Subnet: InvalidSubnet.Conflict: The CIDR '172.16.32.0/24' conflicts with another subnet. Status code: 400`

**Comment by lily-es** [comment]
This also affected me. Would be good to have a fix for this

**Comment by lionello** [comment]
Same issue for me. `pulumi refresh` doesn't help either.

**Comment by GFoley83** [comment]
Does anyone have a workaround that doesn't involve having to nuke the entire stack?

**Comment by danielrbradley** [comment]
This should now be fixed as part of #1153 

We've now added the option to specify the `availabilityZoneCidrMask` which allows you to control the amount of space allocated to each availability zone separate to the number of availability zones you're actually using. 

Example 1: you might initially want to only use 2 AZs but want the option to expand to 3 or 4 AZs in the future. If your VPC is a `/16`, then you can set `availabilityZoneCidrMask` to `18` as it would otherwise default to `17` if only using 2 availability zones.

Example 2: Having created a VPC with 3 AZs but you want to drop down to 2 AZs, you can now remove the last AZ but set the `availabilityZoneCidrMask` to two more than your VPC's CIDR mask. This will delete the third AZ's subnets without affecting the others.

Also, using the new "Auto" subnet strategy, there's is a lot more control available over the specific ordering and spacing of the subnets by using the "Unused" spec type to indicate you specifically want a gap.
